### PR TITLE
Fix some memleaks

### PR DESCRIPTION
**Pull request by The-42**
Had a bit of fun with valgrind and took a stab at fixing a couple memleaks from allocated but unreferenced memory. They're mostly minor in nature except for the ~64 kB crater from XKeyboard initialisation. That one is fixed by commit bf3472e where I'm a bit unsure about whether there could be side effects from using `XCloseDisplay` directly - happy to get a review!

### CODE REVIEW

File: src/enc.h [hunk 1]
```diff
@@ -2092,7 +2092,7 @@ static void doloop (int argc, char *argv[]) {
 			}
 		}
 
-		putenv("ULTRAVNC_DSM_HELPER_LOOP_SET=1");
+		setenv("ULTRAVNC_DSM_HELPER_LOOP_SET", "1", 1);
```
Comment (bk138) [inline src/enc.h]: According to http://www.greenend.org.uk/rjk/tech/putenv.html, setenv is superior, but might not be available. Can you add an autotools check for it and #ifdef accordingly?

File: src/screen.c [hunk 2]
```diff
@@ -3832,12 +3833,16 @@ void announce(int lport, int ssl, char *iface) {
...
+
+		if (host != iface)
+			free(host);
 	}
 }
 
 static void announce_http(int lport, int ssl, char *iface, char *extra) {
 	
 	char *host = this_host();
+	char *phost = host;
```
Comment (The-42) [inline src/screen.c]: ```this_host()``` returns an allocated string which has to be freed explicitly, but depending on certain conditions ```host``` would get either one of the function args or a static string assigned, losing the pointer to the ```this_host()``` result. By using a separate pointer for what gets printed I make sure ```host``` will be freed in any case.

File: src/screen.c [hunk 3]
```diff
@@ -3855,16 +3860,17 @@ static void announce_http(int lport, int ssl, char *iface, char *extra) {
...
+		phost = "localhost";
 	}
-	if (host != NULL) {
+	if (phost != NULL) {
 		if (! inetd) {
-			fprintf(stderr, "%s://%s:%d/%s\n", jvu, host, lport, extra);
+			fprintf(stderr, "%s://%s:%d/%s\n", jvu, phost, lport, extra);
 		}
 	}
+	free(host);
```
Comment (The-42) [inline src/screen.c]: As the function ends there and ```phost``` is a local, I don't see a problem with that.

File: src/xkb_bell.c [hunk 4]
```diff
@@ -76,8 +79,10 @@ void initialize_xkb(void) {
 		return;
 	}
 
-	if (! XkbOpenDisplay(DisplayString(dpy), &xkb_base_event_type, &ir,
-	    NULL, NULL, &reason) ) {
+	xkb_disp = XkbOpenDisplay(DisplayString(dpy), &xkb_base_event_type, &ir,
+		NULL, NULL, &reason);
+
+	if (!xkb_disp) {
```
Comment (The-42) [inline src/xkb_bell.c]: I don't think I'm sure what you mean by that... If ```XkbOpenDisplay``` fails, ```xkb_disp``` is ```NULL``` anyway. The connection needs to stay open until x11vnc is terminated, does it? Or could it be closed right away in any case?

Comment (bk138) [inline src/xkb_bell.c]: Well the connection (in terms of returned pointer) was and is not used anywhere, According to https://linux.die.net/man/3/xkbopendisplay, this cloud maybe be refactored?

Comment (The-42) [inline src/xkb_bell.c]: Oh. Oooooooooh. So keeping the connection open is not relevant for xkeyboard to work. That certainly makes things a bit easier :)

**Comment by bk138** [comment]
disclaimer : I do not know for sure as I did not write the code. we simply have to try out. 

El 9 de febrero de 2017 23:45:39 CET, Bert van Hall <notifications@github.com> escribió:
>The-42 commented on this pull request.
>
>
>
>> @@ -76,8 +79,10 @@ void initialize_xkb(void) {
> 		return;
> 	}
> 
>-	if (! XkbOpenDisplay(DisplayString(dpy), &xkb_base_event_type, &ir,
>-	    NULL, NULL, &reason) ) {
>+	xkb_disp = XkbOpenDisplay(DisplayString(dpy), &xkb_base_event_type,
>&ir,
>+		NULL, NULL, &reason);
>+
>+	if (!xkb_disp) {
>
>Oh. Oooooooooh. So keeping the connection open is not relevant for
>xkeyboard to work. That certainly makes things a bit easier :)
>
>-- 
>You are receiving this because you were assigned.
>Reply to this email directly or view it on GitHub:
>https://github.com/LibVNC/x11vnc/pull/38

-- 
Diese Nachricht wurde von meinem Android-Mobiltelefon mit K-9 Mail gesendet.

**Comment by The-42** [comment]
I built x11vnc from this branch, started it without options. Started `ssvnc`, enabled _Raise on beep_, connected to the x11vnc instance. Executed `x11vnc localhost:0 -remote sendbell` which triggered flashing of the `ssvnc` window and beeps, as it should have. Additionally, the main x11vnc instance logs this:
```
10/02/2017 19:51:09 read X11VNC_REMOTE: cmd=sendbell
10/02/2017 19:51:09 remote_cmd: sendbell.
```

x11vnc 0.9.13 from the repo behaves exactly the same, so I think this works fine without keeping the xkeyboard connection open :)

File: src/xkb_bell.c (hunk 4 above)
Comment (bk138) [inline src/xkb_bell.c]: Can you maybe also ad an explaining comment as this looks quite unusual? ;-)

File: src/screen.c (hunk 2 above)
Comment (bk138) [inline src/screen.c]: Okay, I got it. Pls add a comment to the code explaining that.

File: src/xkb_bell.c (hunk 4 above)
Comment (bk138) [inline src/xkb_bell.c]: maybe this should be refactored to simply use https://linux.die.net/man/3/xkbqueryextension @The-42 ?

File: src/enc.h [hunk 5]
```diff
@@ -2092,7 +2092,7 @@ static void doloop (int argc, char *argv[]) {
 			}
```
Comment (bk138) [inline src/enc.h]: i've cherry-picked this one already @The-42

File: src/xkb_bell.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (bk138) [reply src/xkb_bell.c]: Fixed with 093a25114926285c0cbd78c5e9a2a8e91714536d

**Pull request closed without merging**